extensions/amp-access/0.1/amp-access.js

```diff
@@ -145,30 +152,52 @@ export class AccessService {
   }
 
   /**
+   * @return {!Promise<string>}
+   * @private
+   */
+  getReaderId_() {
+    if (!this.readerIdPromise_) {
+      // TODO(dvoytenko): Confirm no consent is needed. The assumtpion is that
+      // this is the case given that this is an essential part of access system.
+      const consent = Promise.resolve();
+      this.readerIdPromise_ = this.cid_.get('amp-access', consent);
+    }
+    return this.readerIdPromise_;
+  }
+
+  /**
+   * @param {string} url
+   * @return {!Promise<string>}
+   * @private
+   */
+  buildUrl_(url) {
+    return this.getReaderId_().then(readerId => {
+      return this.urlReplacements_.expand(url, {
+        'READER_ID': readerId
+      });
+    });
+  }
+
+  /**
    * @return {!Promise}
    * @private
    */
   runAuthorization_() {
     log.fine(TAG, 'Start authorization via ', this.config_.authorization);
     this.toggleTopClass_('amp-access-loading', true);
-
-    // TODO(dvoytenko): produce READER_ID and create the URL substition for it.
-    return this.urlReplacements_.expand(this.config_.authorization)
-        .then(url => {
-          log.fine(TAG, 'Authorization URL: ', url);
-          return this.xhr_.fetchJson(url, {credentials: 'include'});
-        })
-        .then(response => {
-          log.fine(TAG, 'Authorization response: ', response);
-          this.toggleTopClass_('amp-access-loading', false);
-          onDocumentReady(this.win.document, () => {
-            this.applyAuthorization_(response);
-          });
-        })
-        .catch(error => {
-          log.error(TAG, 'Authorization failed: ', error);
-          this.toggleTopClass_('amp-access-loading', false);
-        });
+    return this.buildUrl_(this.config_.authorization).then(url => {
+      log.fine(TAG, 'Authorization URL: ', url);
+      return this.xhr_.fetchJson(url, {credentials: 'include'});
+    }).then(response => {
+      log.fine(TAG, 'Authorization response: ', response);
+      this.toggleTopClass_('amp-access-loading', false);
```

Comment:
I want to start XHR asap. However, since the follow up code scans DOM, I have to wait for DOMReady.